src/App.tsx

```diff
@@ -1,24 +1,301 @@
+import { Errors } from './types/Errors';
+import { FilterOptions } from './types/FilterOptions';
 
-const USER_ID = 0;
+const USER_ID = 8;
+
+const getFilteredTodos = (todos: Todo[], filterBy: FilterOptions) => {
+  switch (filterBy) {
+    case FilterOptions.Active:
+      return todos.filter(todo => !todo.completed);
+    case FilterOptions.Completed:
+      return todos.filter(todo => todo.completed);
+    default:
+      return todos;
+  }
+};
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterBy, setFilterBy] = useState(FilterOptions.All);
+  const filteredTodos = getFilteredTodos(todos, filterBy);
+  const uncompletedTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+  const [loadingTodosId, setLoadingTodosId] = useState<number[]>([]);
+  const [temporaryTodo, setTemporaryTodo] = useState<Todo | null>(null);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [errorMasage, setErrorMasage] = useState(Errors.allGood);
+  const titleField = useRef<HTMLInputElement>(null);
+  const [isDisabled, setIsDisabled] = useState(false);
+
+  const reset = () => {
+    setTodoTitle('');
+    setErrorMasage(Errors.allGood);
+    setTemporaryTodo(null);
+    setIsDisabled(false);
+    titleField.current?.focus();
+  };
+
+  function loadTodos() {
+    clientService.getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => setErrorMasage(Errors.cantGetArray));
+  }
+
+  const addProcessing = (id: number) => {
+    setLoadingTodosId(current => [...current, id]);
+  };
+
+  const removeProcessing = (idToRemove: number) => {
+    setLoadingTodosId(current => current.filter(id => id !== idToRemove));
+  };
+
+  function updateTodoOnServer(todoForChange: Todo, changedTodo: Todo) {
+    setTodos(currentTodos => {
+      const newTodos = [...currentTodos];
+      const index = newTodos.findIndex(
+        currentTodo => currentTodo.id === changedTodo.id,
+      );
+
+      newTodos.splice(index, 1, todoForChange);
+
+      return newTodos;
+    });
+  }
+
+  const addTodo = (
+    title: string, userId: number, completed: boolean,
+  ) => {
+    setIsDisabled(true);
+    setTemporaryTodo({
+      title, userId, completed: false, id: 0,
+    });
+    addProcessing(0);
+
+    clientService.addTodo({ title, userId, completed })
+      .then(newTodo => {
+        setTodos(currentTodos => {
+          return [...currentTodos, newTodo];
+        });
+        reset();
+        removeProcessing(0);
+      })
+      .catch(() => {
+        setErrorMasage(Errors.cantAdd);
+        setTemporaryTodo(null);
+        setIsDisabled(false);
+        setLoadingTodosId([]);
+      });
+  };
+
+  const updateTodo = (updatedTodo: Todo) => {
+    addProcessing(updatedTodo.id);
+
+    return clientService.updateTodo(updatedTodo)
+      .catch((e) => {
+        setErrorMasage(Errors.cantUpdate);
+        removeProcessing(updatedTodo.id);
+        throw e;
```

Comment:
do not throw an error here - you handled it already, so no need to throw it again. It will lead to a page crash (because you don't try to catch the error thrown here somewhere else)